File: apps/app-4/frontend/src/api/health-profile.ts
import { axiosInstance } from './axios';

// Types
export interface HealthProfile {
  id: number;
  user_id: number;
  full_name: string;
  birth_date?: string;
  gender?: string;
  blood_type?: string;
  height_cm: number;
  weight_kg: number;
  bmi: number;
  body_fat_percent?: number;
  last_weight_date?: string;
  overall_score: number;
  sleep_score: number;
  nutrition_score: number;
  activity_score: number;
  hydration_score: number;
  mental_score: number;
  water_target_ml: number;
  steps_target: number;
  sleep_target_hours: number;
  calories_target: number;
  chronic_conditions: string;
  allergies: string;
  family_history: string;
  surgeries: string;
  smoking_status: string;
  alcohol_use: string;
  exercise_level: string;
  diet_type: string;
  health_goals: string;
  last_triage_id?: number;
  last_triage_date?: string;
  profile_completeness: number;
  created_at: string;
  updated_at: string;
}

export interface DailyCheckIn {
  id: number;
  user_id: number;
  date: string;
  sleep_hours: number;
  sleep_quality: number;
  woke_up_time: string;
  bed_time: string;
  steps: number;
  active_minutes: number;
  exercise_type?: string;
  calories_burned: number;
  calories_consumed: number;
  water_ml: number;
  meals_count: number;
  meals_log: string;
  weight_kg?: number;
  blood_pressure_sys?: number;
  blood_pressure_dia?: number;
  heart_rate?: number;
  blood_glucose?: number;
  mood_score: number;
  energy_level: number;
  stress_level: number;
  symptoms: string;
  notes: string;
  ai_insights: string;
  ai_recommendations: string;
  risk_flags: string;
  source: string;
  created_at: string;
}

export interface HealthMetric {
  id: number;
  user_id: number;
  type: string;
  value: number;
  unit: string;
  date: string;
  source: string;
  created_at: string;
}

export interface Medication {
  id: number;
  user_id: number;
  name: string;
  dosage: string;
  frequency: string;
  times: string;
  start_date: string;
  end_date?: string;
  instructions: string;
  prescribed_by?: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface MedicationLog {
  id: number;
  medication_id: number;
  user_id: number;
  taken_at: string;
  scheduled_for: string;
  status: string;
  notes?: string;
}

export interface Vaccine {
  id: number;
  user_id: number;
  name: string;
  manufacturer?: string;
  batch?: string;
  dose_number: number;
  total_doses: number;
  applied_at: string;
  next_dose_at?: string;
  location?: string;
  applied_by?: string;
  notes?: string;
  created_at: string;
}

export interface Exam {
  id: number;
  user_id: number;
  name: string;
  type: string;
  requested_by?: number;
  requested_at?: string;
  scheduled_at?: string;
  completed_at?: string;
  result_url?: string;
  results?: string;
  status: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface HealthGoal {
  id: number;
  user_id: number;
  title: string;
  description: string;
  category: string;
  target_value: number;
  current_value: number;
  unit: string;
  start_date: string;
  target_date: string;
  status: string;
  progress: number;
  set_by: string;
  created_at: string;
  updated_at: string;
}

export interface Achievement {
  id: number;
  user_id: number;
  type: string;
  name: string;
  description: string;
  icon: string;
  earned_at: string;
  points: number;
}

export interface HealthSummary {
  profile: HealthProfile;
  todayCheckIn?: DailyCheckIn;
  healthScore: number;
  healthTrend: 'up' | 'down' | 'stable';
  sleepScore: number;
  hydrationScore: number;
  activityScore: number;
  nutritionScore: number;
  mentalScore: number;
  pendingMedications: number;
  pendingExams: number;
  overdueVaccines: number;
  activeGoals: number;
  achievements: number;
  profileCompleteness: number;
  alerts: { type: string; message: string }[];
}


// API Functions

// Profile
export async function getMyProfile(): Promise<HealthProfile> {
  const response = await axiosInstance.get('/health-profile/me');
  return response.data;
}

export async function updateMyProfile(profile: Partial<HealthProfile>): Promise<HealthProfile> {
  const response = await axiosInstance.put('/health-profile/me', profile);
  return response.data;
}

export async function getHealthSummary(): Promise<HealthSummary> {
  const response = await axiosInstance.get('/health-profile/summary');
  return response.data;
}

export async function updateProfileFromTriage(triageId: number): Promise<void> {
  await axiosInstance.post(`/health-profile/from-triage/${triageId}`);
}

// Daily Check-ins
export async function createDailyCheckIn(checkIn: Partial<DailyCheckIn>): Promise<DailyCheckIn> {
  const response = await axiosInstance.post('/health-profile/check-in', checkIn);
  return response.data;
}

export async function getTodayCheckIn(): Promise<DailyCheckIn | null> {
  try {
    const response = await axiosInstance.get('/health-profile/check-in/today');
    return response.data;
  } catch {
    return null;
  }
}

export async function getCheckInHistory(days: number = 30): Promise<DailyCheckIn[]> {
  const response = await axiosInstance.get(`/health-profile/check-in/history?days=${days}`);
  return response.data || [];
}

export async function processAIChatCheckIn(message: string): Promise<DailyCheckIn> {
  const response = await axiosInstance.post('/health-profile/check-in/ai', { message });
  return response.data;
}

// Metrics
export async function recordMetric(metric: Partial<HealthMetric>): Promise<HealthMetric> {
  const response = await axiosInstance.post('/health-profile/metrics', metric);
  return response.data;
}

export async function getMetrics(type?: string, days: number = 30): Promise<HealthMetric[]> {
  const params = new URLSearchParams({ days: days.toString() });
  if (type) params.append('type', type);
  const response = await axiosInstance.get(`/health-profile/metrics?${params}`);
  return response.data || [];
}

// Medications
export async function getMedications(): Promise<Medication[]> {
  const response = await axiosInstance.get('/health-profile/medications');
  return response.data || [];
}

export async function createMedication(medication: Partial<Medication>): Promise<Medication> {
  const response = await axiosInstance.post('/health-profile/medications', medication);
  return response.data;
}

export async function logMedication(log: Partial<MedicationLog>): Promise<MedicationLog> {
  const response = await axiosInstance.post('/health-profile/medications/log', log);
  return response.data;
}

// Vaccines
export async function getVaccines(): Promise<Vaccine[]> {
  const response = await axiosInstance.get('/health-profile/vaccines');
  return response.data || [];
}

export async function createVaccine(vaccine: Partial<Vaccine>): Promise<Vaccine> {
  const response = await axiosInstance.post('/health-profile/vaccines', vaccine);
  return response.data;
}

// Exams
export async function getExams(): Promise<Exam[]> {
  const response = await axiosInstance.get('/health-profile/exams');
  return response.data || [];
}

export async function createExam(exam: Partial<Exam>): Promise<Exam> {
  const response = await axiosInstance.post('/health-profile/exams', exam);
  return response.data;
}

// Goals
export async function getHealthGoals(): Promise<HealthGoal[]> {
  const response = await axiosInstance.get('/health-profile/goals');
  return response.data || [];
}

export async function createHealthGoal(goal: Partial<HealthGoal>): Promise<HealthGoal> {
  const response = await axiosInstance.post('/health-profile/goals', goal);
  return response.data;
}

export async function updateHealthGoal(id: number, goal: Partial<HealthGoal>): Promise<HealthGoal> {
  const response = await axiosInstance.put(`/health-profile/goals/${id}`, goal);
  return response.data;
}

// Achievements
export async function getAchievements(): Promise<Achievement[]> {
  const response = await axiosInstance.get('/health-profile/achievements');
  return response.data || [];
}

// Helper functions
export function parseJsonArray(jsonString: string): string[] {
  if (!jsonString || jsonString === '[]') return [];
  try {
    return JSON.parse(jsonString);
  } catch {
    return [];
  }
}

export function calculateAge(birthDate: string): number {
  const birth = new Date(birthDate);
  const today = new Date();
  let age = today.getFullYear() - birth.getFullYear();
  const monthDiff = today.getMonth() - birth.getMonth();
  if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birth.getDate())) {
    age--;
  }
  return age;
}

export function getScoreColor(score: number): string {
  if (score >= 80) return 'text-emerald-500';
  if (score >= 60) return 'text-yellow-500';
  if (score >= 40) return 'text-orange-500';
  return 'text-red-500';
}

export function getScoreGradient(score: number): string {
  if (score >= 80) return 'from-emerald-500 to-teal-500';
  if (score >= 60) return 'from-yellow-500 to-amber-500';
  if (score >= 40) return 'from-orange-500 to-red-400';
  return 'from-red-500 to-rose-600';
}

export function getScoreLabel(score: number): string {
  if (score >= 80) return 'Excelente';
  if (score >= 60) return 'Bom';
  if (score >= 40) return 'Regular';
  return 'Precisa melhorar';
}

File: apps/app-4/backend/internal/core/domain/health_profile.go
package domain

import (
	"time"
)

// HealthProfile represents the complete health profile of a patient
// This is the central data structure that connects all health data
type HealthProfile struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"uniqueIndex"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	// Basic Info (extracted from triage/conversations)
	FullName    string     `json:"full_name"`
	BirthDate   *time.Time `json:"birth_date,omitempty"`
	Gender      string     `json:"gender,omitempty"`      // male, female, other, prefer_not_say
	BloodType   string     `json:"blood_type,omitempty"`  // A+, A-, B+, B-, AB+, AB-, O+, O-

	// Anthropometric Data
	HeightCm       float64    `json:"height_cm"`
	WeightKg       float64    `json:"weight_kg"`
	BMI            float64    `json:"bmi"`
	BodyFatPercent *float64   `json:"body_fat_percent,omitempty"`
	LastWeightDate *time.Time `json:"last_weight_date,omitempty"`

	// Health Scores (0-100)
	OverallScore    int `json:"overall_score" gorm:"default:50"`
	SleepScore      int `json:"sleep_score" gorm:"default:50"`
	NutritionScore  int `json:"nutrition_score" gorm:"default:50"`
	ActivityScore   int `json:"activity_score" gorm:"default:50"`
	HydrationScore  int `json:"hydration_score" gorm:"default:50"`
	MentalScore     int `json:"mental_score" gorm:"default:50"`

	// Daily Targets
	WaterTargetMl     int `json:"water_target_ml" gorm:"default:2000"`
	StepsTarget       int `json:"steps_target" gorm:"default:10000"`
	SleepTargetHours  int `json:"sleep_target_hours" gorm:"default:8"`
	CaloriesTarget    int `json:"calories_target" gorm:"default:2000"`

	// Medical History (JSON arrays)
	ChronicConditions string `json:"chronic_conditions"` // JSON array
	Allergies         string `json:"allergies"`          // JSON array
	FamilyHistory     string `json:"family_history"`     // JSON array
	Surgeries         string `json:"surgeries"`          // JSON array

	// Lifestyle
	SmokingStatus   string `json:"smoking_status"`   // never, former, current
	AlcoholUse      string `json:"alcohol_use"`      // never, occasional, moderate, heavy
	ExerciseLevel   string `json:"exercise_level"`   // sedentary, light, moderate, active, very_active
	DietType        string `json:"diet_type"`        // omnivore, vegetarian, vegan, pescatarian, other

	// Goals (JSON)
	HealthGoals string `json:"health_goals"` // JSON array of goals

	// Last Triage Data
	LastTriageID   *uint      `json:"last_triage_id,omitempty"`
	LastTriageDate *time.Time `json:"last_triage_date,omitempty"`

	// Profile Completeness (0-100)
	ProfileCompleteness int `json:"profile_completeness" gorm:"default:0"`

	// Relations
	User User `json:"user,omitempty" gorm:"foreignKey:UserID"`
}

// DailyCheckIn represents a daily health check-in from the user
type DailyCheckIn struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"index"`
	Date      time.Time `json:"date" gorm:"index"`
	CreatedAt time.Time `json:"created_at"`

	// Sleep
	SleepHours   float64 `json:"sleep_hours"`
	SleepQuality int     `json:"sleep_quality"` // 1-5
	WokeUpTime   string  `json:"woke_up_time"`
	BedTime      string  `json:"bed_time"`

	// Activity
	Steps          int     `json:"steps"`
	ActiveMinutes  int     `json:"active_minutes"`
	ExerciseType   string  `json:"exercise_type,omitempty"`
	CaloriesBurned int     `json:"calories_burned"`

	// Nutrition
	CaloriesConsumed int    `json:"calories_consumed"`
	WaterMl          int    `json:"water_ml"`
	MealsCount       int    `json:"meals_count"`
	MealsLog         string `json:"meals_log"` // JSON array of meals

	// Vitals (if measured)
	WeightKg         *float64 `json:"weight_kg,omitempty"`
	BloodPressureSys *int     `json:"blood_pressure_sys,omitempty"`
	BloodPressureDia *int     `json:"blood_pressure_dia,omitempty"`
	HeartRate        *int     `json:"heart_rate,omitempty"`
	BloodGlucose     *float64 `json:"blood_glucose,omitempty"`

	// Mood & Symptoms
	MoodScore    int    `json:"mood_score"` // 1-5
	EnergyLevel  int    `json:"energy_level"` // 1-5
	StressLevel  int    `json:"stress_level"` // 1-5
	Symptoms     string `json:"symptoms"`     // JSON array
	Notes        string `json:"notes"`

	// AI Analysis
	AIInsights     string `json:"ai_insights"`      // AI-generated insights
	AIRecommendations string `json:"ai_recommendations"` // AI recommendations
	RiskFlags      string `json:"risk_flags"`       // JSON array of detected risks

	// Source
	Source string `json:"source"` // manual, triage, sensor, ai_chat
}

// HealthMetric represents a single health metric measurement
type HealthMetric struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"index"`
	Type      string    `json:"type" gorm:"index"` // weight, steps, water, sleep, calories, heart_rate, etc.
	Value     float64   `json:"value"`
	Unit      string    `json:"unit"`
	Date      time.Time `json:"date" gorm:"index"`
	Source    string    `json:"source"` // manual, triage, sensor, ai_chat
	CreatedAt time.Time `json:"created_at"`
}

// Medication represents a medication the user is taking
type Medication struct {
	ID          uint       `json:"id" gorm:"primaryKey"`
	UserID      uint       `json:"user_id" gorm:"index"`
	Name        string     `json:"name"`
	Dosage      string     `json:"dosage"`
	Frequency   string     `json:"frequency"` // daily, twice_daily, weekly, as_needed
	Times       string     `json:"times"`     // JSON array of times ["08:00", "20:00"]
	StartDate   time.Time  `json:"start_date"`
	EndDate     *time.Time `json:"end_date,omitempty"`
	Instructions string    `json:"instructions"`
	PrescribedBy *uint     `json:"prescribed_by,omitempty"` // Doctor ID
	IsActive    bool       `json:"is_active" gorm:"default:true"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

// MedicationLog represents when a user took their medication
type MedicationLog struct {
	ID           uint      `json:"id" gorm:"primaryKey"`
	MedicationID uint      `json:"medication_id" gorm:"index"`
	UserID       uint      `json:"user_id" gorm:"index"`
	TakenAt      time.Time `json:"taken_at"`
	ScheduledFor time.Time `json:"scheduled_for"`
	Status       string    `json:"status"` // taken, missed, skipped
	Notes        string    `json:"notes,omitempty"`
}

// Vaccine represents a vaccine record
type Vaccine struct {
	ID           uint       `json:"id" gorm:"primaryKey"`
	UserID       uint       `json:"user_id" gorm:"index"`
	Name         string     `json:"name"`
	Manufacturer string     `json:"manufacturer,omitempty"`
	Batch        string     `json:"batch,omitempty"`
	DoseNumber   int        `json:"dose_number"`
	TotalDoses   int        `json:"total_doses"`
	AppliedAt    time.Time  `json:"applied_at"`
	NextDoseAt   *time.Time `json:"next_dose_at,omitempty"`
	Location     string     `json:"location,omitempty"`
	AppliedBy    string     `json:"applied_by,omitempty"`
	Notes        string     `json:"notes,omitempty"`
	CreatedAt    time.Time  `json:"created_at"`
}

// Exam represents a medical exam record
type Exam struct {
	ID          uint       `json:"id" gorm:"primaryKey"`
	UserID      uint       `json:"user_id" gorm:"index"`
	Name        string     `json:"name"`
	Type        string     `json:"type"` // blood, imaging, other
	RequestedBy *uint      `json:"requested_by,omitempty"` // Doctor ID
	RequestedAt *time.Time `json:"requested_at,omitempty"`
	ScheduledAt *time.Time `json:"scheduled_at,omitempty"`
	CompletedAt *time.Time `json:"completed_at,omitempty"`
	ResultURL   string     `json:"result_url,omitempty"`
	Results     string     `json:"results,omitempty"` // JSON or text
	Status      string     `json:"status"`            // pending, scheduled, completed, cancelled
	Notes       string     `json:"notes,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

// HealthGoal represents a health goal set by user or AI
type HealthGoal struct {
	ID          uint       `json:"id" gorm:"primaryKey"`
	UserID      uint       `json:"user_id" gorm:"index"`
	Title       string     `json:"title"`
	Description string     `json:"description"`
	Category    string     `json:"category"` // weight, activity, nutrition, sleep, medication, other
	TargetValue float64    `json:"target_value"`
	CurrentValue float64   `json:"current_value"`
	Unit        string     `json:"unit"`
	StartDate   time.Time  `json:"start_date"`
	TargetDate  time.Time  `json:"target_date"`
	Status      string     `json:"status"` // active, completed, abandoned
	Progress    int        `json:"progress"` // 0-100
	SetBy       string     `json:"set_by"`   // user, ai, doctor
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

// Achievement represents a health achievement/badge
type Achievement struct {
	ID          uint      `json:"id" gorm:"primaryKey"`
	UserID      uint      `json:"user_id" gorm:"index"`
	Type        string    `json:"type"`        // streak, milestone, challenge
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Icon        string    `json:"icon"`
	EarnedAt    time.Time `json:"earned_at"`
	Points      int       `json:"points"`
}

// TriageExtractedData represents structured data extracted from triage by AI
type TriageExtractedData struct {
	// Personal Info
	Age              *int     `json:"age,omitempty"`
	Gender           string   `json:"gender,omitempty"`
	WeightKg         *float64 `json:"weight_kg,omitempty"`
	HeightCm         *float64 `json:"height_cm,omitempty"`

	// Symptoms & Conditions
	CurrentSymptoms  []string `json:"current_symptoms"`
	ChronicConditions []string `json:"chronic_conditions"`
	Allergies        []string `json:"allergies"`
	
	// Medications
	CurrentMedications []ExtractedMedication `json:"current_medications"`
	
	// Lifestyle
	SmokingStatus    string   `json:"smoking_status"`
	AlcoholUse       string   `json:"alcohol_use"`
	ExerciseLevel    string   `json:"exercise_level"`
	SleepHours       *float64 `json:"sleep_hours,omitempty"`
	
	// Vitals mentioned
	BloodPressure    string   `json:"blood_pressure,omitempty"`
	HeartRate        *int     `json:"heart_rate,omitempty"`
	
	// Goals mentioned
	HealthGoals      []string `json:"health_goals"`
	
	// Risk factors
	RiskFactors      []string `json:"risk_factors"`
}

type ExtractedMedication struct {
	Name     string `json:"name"`
	Dosage   string `json:"dosage"`
	Frequency string `json:"frequency"`
}
